server/genquery2/src/genquery2_sql.cpp

```diff
@@ -0,0 +1,1172 @@
+        }
+
+        return graph;
+    } // init_graph
+
+    auto generate_inner_joins(const graph_type& _graph,
+                              const std::vector<std::string_view>& _tables,
+                              const std::map<std::string, std::string>& _table_aliases) -> std::vector<std::string>
+    {
+        const auto get_table_join = [&_graph, &_table_aliases](const auto& _t1, const auto& _t2) -> std::string {
+            auto t1_idx = to_index(_t1);
+            auto t2_idx = to_index(_t2);
+            const auto [edge, exists] = boost::edge(t1_idx, t2_idx, _graph);
+
+            if (!exists) {
+                return {};
+            }
+
+            std::string_view t2_alias = _table_aliases.at(std::string{_t2});
+            const auto sql = fmt::format("inner join {} {} on {}", _t2, t2_alias, _graph[edge].join_condition);
+
+            // It is likely that the order of the tables passed do NOT match the order of the join expression.
+            // To resolve this, each edge property contains its position in the table_edges array. This is used
+            // to lookup the edge definition in table_edges. This allows the parser to reorder the tables to
+            // satisfy the table join expression.
+            const auto& edge_def =
+                table_edges[_graph[edge].position]; // NOLINT(cppcoreguidelines-pro-bounds-constant-array-index)
+            auto& t1_alias = _table_aliases.at(
+                table_names[edge_def.first]); // NOLINT(cppcoreguidelines-pro-bounds-constant-array-index)
+            t2_alias = _table_aliases.at(
+                table_names[edge_def.second]); // NOLINT(cppcoreguidelines-pro-bounds-constant-array-index)
+
+            return fmt::format(fmt::runtime(sql), t1_alias, t2_alias);
+        };
+
+        // The order of inner joins may or may not matter depending on the design of your SQL database.
+        // There are cases in which the order of the inner joins matter in iRODS (e.g. permissions).
+        // The following algorithm resolves the joins by scanning the list of tables multiple times.
+        // The list of tables is obtained during parsing of the string.
+
+        std::vector<std::string> inner_joins;
+        inner_joins.reserve(_tables.size() - 1);
+
+        // Copy all entries from "_tables" into the list except the very first one.
+        // The first element is the table we are trying to join to. So, we consider that one handled.
+        std::vector<std::string> remaining{std::begin(_tables) + 1, std::end(_tables)};
+        log_gq::debug(fmt::format("remaining = [{}]", fmt::join(remaining, ", ")));
+
+        std::vector<std::string> processed;
+        processed.reserve(_tables.size());
+        processed.emplace_back(std::string{_tables.front()});
+        log_gq::debug(fmt::format("processed = [{}]", fmt::join(processed, ", ")));
+
+        for (decltype(_tables.size()) i = 0; i < _tables.size() - 1; ++i) {
```

Comment:
Any reason this isn't a range-based for loop? Doesn't have to be, just curious.

Reply:
Range-based for loops require iterating over all elements which this does not.

There may be other reasons for using this, but I can't tell due to GitHub showing only a few lines from this page.

Reply:
Yeah, `_tables` and `i` aren't being used anyway, so it doesn't really matter. Was just wondering. I'll resolve this.